roachprod: add command to download the latest pprof file

nameisbhaskar opened the pull request:
Currently, whenever we see a spike in the CPU for a specific cluster, we have to download the latest pprof file for analysis. We generally select the latest file across all nodes or a specific set of nodes. This becomes a very manual process as we need to check each cluster to find the suitable file to download.
This command helps in finding the latest pprof to download across all nodes or given nodes. The command also takes a time parameter as input with which we can download the latest pprof before a specific time.

Epic: none
Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/131567)
<!-- Reviewable:end -->

nameisbhaskar committed:
roachprod: add command to download the latest pprof file

Currently, whenever we see a spike in the CPU for a specific cluster, we have to download the latest pprof file for analysis. We generally select the latest file across all nodes or a specific set of nodes. This becomes a very manual process as we need to check each cluster to find the suitable file to download.
This command helps in finding the latest pprof to download across all nodes or given nodes. The command also takes a time parameter as input with which we can download the latest pprof before a specific time.

Epic: none
Release note: None

srosenberg commented on pkg/cmd/roachprod/flags.go:
```diff
@@ -490,4 +491,6 @@ func initFlags() {
 	fetchLogsCmd.Flags().DurationVarP(&fetchLogsTimeout,
 		"timeout", "t", 5*time.Minute, "Timeout for fetching the logs from the cluster nodes")
+	pprofTimeBefore = getLatestPprof.Flags().StringP(
+		"time-before", "t", "", "Timeout for fetching the logs from the cluster nodes")
```
Forgot to update the description? I'd also use `Duration` or at least hint at what time units are expected.

srosenberg commented on pkg/cmd/roachprod/main.go:
```diff
@@ -2109,3 +2110,19 @@ The logs will be placed in the directory if specified or in the directory named
+
+var getLatestPprof = &cobra.Command{
+	Use:   "get-latest-pprof <cluster> [flags]",
+	Short: "downloads the latest pprof file before the provided time.",
```
Can we be more precise on what "before" actually means? Is it `<= now` or `< now`?

herkolategan commented on pkg/cmd/roachprod/main.go:
```diff
@@ -2109,3 +2110,19 @@ The logs will be placed in the directory if specified or in the directory named
 	}),
 }
+
+var getLatestPprof = &cobra.Command{
```
Nit: For capitalisation consistency when `pprof` >= 2nd word, I noticed `workloadPProfStartPort`. Just so we don't have a mix of `Pprof` and `PProf` in the code.

herkolategan commented on pkg/roachprod/roachprod_pprof.go:
```diff
@@ -0,0 +1,202 @@
+
+// getLatestPprofInfo gets the latest pprofInfo across all clusters before beforeTime
+// it returns nil if no pprof were found.
+func getLatestPprofInfo(
```
Nit: Same as above `DownloadLatestPProfFile` vs. `getLatestPprofInfo` (prefer `PProf`)

herkolategan commented on pkg/cmd/roachprod/flags.go:
```diff
@@ -490,4 +491,6 @@ func initFlags() {
 		"time-range", []int64{}, "grafana annotation time range in epoch time")
 	fetchLogsCmd.Flags().DurationVarP(&fetchLogsTimeout,
 		"timeout", "t", 5*time.Minute, "Timeout for fetching the logs from the cluster nodes")
+	pprofTimeBefore = getLatestPProf.Flags().StringP(
```
Question around the flag: Is there a specific reason we deviate from using the `Var` variant used throughout for flags and not passing a &(str) pointer for `pprofTimeBefore`?

nameisbhaskar replied:
I wanted to use the shorthand which was not available for StringVar

nameisbhaskar commented on pkg/cmd/roachprod/flags.go:
```diff
@@ -490,4 +491,6 @@ func initFlags() {
 		"time-range", []int64{}, "grafana annotation time range in epoch time")
 	fetchLogsCmd.Flags().DurationVarP(&fetchLogsTimeout,
 		"timeout", "t", 5*time.Minute, "Timeout for fetching the logs from the cluster nodes")
+	pprofTimeBefore = getLatestPProf.Flags().StringP(
```
I am fine either ways. Let me make it a arg to make it very explicit.

nameisbhaskar commented:
Thanks @srosenberg @herkolategan @DarrylWong for reviewing the changes. Sorry for the delay in reverting back to the changes as I got occupied with the 150 node cluster tasks. I have addressed your comments. PTAL.

nameisbhaskar commented:
Thanks @herkolategan for approving the changes!

nameisbhaskar commented:
bors r=@herkolategan
